Detectors/ZDC/simulation/src/Digits2Raw.cxx

```diff
@@ -0,0 +1,874 @@
+#include <string>
+#include <TFile.h>
+#include <TTree.h>
+#include <TRandom.h>
+#include <TMath.h>
+#include "SimulationDataFormat/MCTruthContainer.h"
+#include "SimulationDataFormat/MCCompLabel.h"
+#include "ZDCBase/Constants.h"
+#include "ZDCBase/ModuleConfig.h"
+#include "ZDCSimulation/Digitizer.h"
+#include "DataFormatsZDC/BCData.h"
+#include "DataFormatsZDC/ChannelData.h"
+#include "ZDCSimulation/Digits2Raw.h"
+#include "ZDCSimulation/MCLabel.h"
+#include "ZDCSimulation/ZDCSimParam.h"
+#include "CommonUtils/StringUtils.h"
+#include "FairLogger.h"
+
+using namespace o2::zdc;
+
+//ClassImp(Digits2Raw);
+//______________________________________________________________________________
+void Digits2Raw::processDigits(const std::string& outDir, const std::string& fileDigitsName)
+{
+  auto& sopt = ZDCSimParam::Instance();
+  mIsContinuous = sopt.continuous;
+
+  if (!mModuleConfig) {
+    LOG(ERROR) << "Missing ModuleConfig configuration object";
+    return;
+  }
+
+  if (!mSimCondition) {
+    LOG(ERROR) << "Missing SimCondition configuration object";
+    return;
+  }
+
+  if (mNEmpty < 0) {
+    LOG(ERROR) << "Bunch crossing map is not initialized";
+    return;
+  }
+
+  if (mNEmpty == 0) {
+    LOG(WARNING) << "Bunch crossing map has zero clean empty bunches";
+  }
+
+  setTriggerMask();
+
+  std::string outd = outDir;
+  if (outd.back() != '/') {
+    outd += '/';
+  }
+
+  mLinkID = uint32_t(0);
+  mCruID = uint16_t(0);
+  mEndPointID = uint32_t(0);
+  for (int ilink = 0; ilink < NLinks; ilink++) {
+    mFeeID = uint64_t(ilink);
+    std::string outFileLink = mOutputPerLink ? o2::utils::concat_string(outd, "zdc_link", std::to_string(ilink), ".raw") : o2::utils::concat_string(outd, "zdc.raw");
+    mWriter.registerLink(mFeeID, mCruID, mLinkID, mEndPointID, outFileLink);
+  }
+
+  std::unique_ptr<TFile> digiFile(TFile::Open(fileDigitsName.c_str()));
+  if (!digiFile || digiFile->IsZombie()) {
+    LOG(ERROR) << "Failed to open input digits file " << fileDigitsName;
+    return;
+  }
+
+  TTree* digiTree = (TTree*)digiFile->Get("o2sim");
+  if (!digiTree) {
+    LOG(ERROR) << "Failed to get digits tree";
+    return;
+  }
+
+  o2::dataformats::MCTruthContainer<o2::zdc::MCLabel>* labelsPtr = nullptr;
+
+  if (digiTree->GetBranch("ZDCDigitBC")) {
+    digiTree->SetBranchAddress("ZDCDigitBC", &mzdcBCDataPtr);
+  } else {
+    LOG(ERROR) << "Branch ZDCDigitBC is missing";
+    return;
+  }
+
+  if (digiTree->GetBranch("ZDCDigitCh")) {
+    digiTree->SetBranchAddress("ZDCDigitCh", &mzdcChDataPtr);
+  } else {
+    LOG(ERROR) << "Branch ZDCDigitCh is missing";
+    return;
+  }
+
+  if (digiTree->GetBranch("ZDCDigitLabels")) {
+    digiTree->SetBranchAddress("ZDCDigitLabels", &labelsPtr);
+    LOG(INFO) << "Branch ZDCDigitLabels is connected";
+  } else {
+    LOG(INFO) << "Branch ZDCDigitLabels is missing";
+  }
```

Comment:
why do you need to read lables? You can even suppress the branch to avoid root reading it automatically: ``digiTree->SetBranchStatus("ZDCDigitLabel*", 0);``